### DISCUSSION TITLE
SwiftUI implementation of Compose Children widget

### DISCUSSION THREAD

**Discussion by plusmobileapps** [badoo/Decompose#264] [Ideas]
When using the [`Children`](https://github.com/arkivanov/Decompose/blob/master/extensions-compose-jetpack/src/main/java/com/arkivanov/decompose/extensions/compose/jetpack/Children.kt) widget in Jetpack Compose, it is a lot easier to implement an Android app navigation transitions than iOS. It would be nice to have a SwiftUI equivalent to provide a similar animation api using the Swift UI transition API like in the [todo app](https://github.com/JetBrains/compose-jb/blob/master/examples/todoapp/ios/ios/RootView.swift#L21). 

The problem currently in the SwiftUI sample would be to add a third page after the edit with another transition sliding in, this causes the edit page to slide out to the right when the third page is coming in from the right. While on Android this animation would have the edit page slide to the left while the new page is coming in from the right. 

I'm not really a Swift developer, but was wondering if using a library like [NavigationStack](https://github.com/indieSoftware/NavigationStack) could be used to integrate with Decompose's navigation API to perform the transitions based off the states being emitted by the router. Or perhaps do something similar to implementing a [native back swipe suggestion](https://github.com/arkivanov/Decompose/discussions/203#discussioncomment-1162765) with a `ZStack` view that can have minimal logic to handle the transitioning?

**Comment by arkivanov** [comment]
Thanks for he question! This is partially related to #228, where I explained why it's not possible to create and publish a companion Swift library. So the only reasonable way currently is to provide copy-paste solutions.

A while ago, before the todo-app example was moved from this repository to compose-jb, I created a [RouterView](https://github.com/arkivanov/Decompose/blob/fbec0ac19f813bb6e619ceb6d482c098e8e1dfc9/sample/todo/ios-app/ios-app/RouterView.swift) for SwiftUI. While moving the example to another repository, I decided to avoid the RouterVIew for simplicity. The `RouterView` in its original implementation does not have any animations, but it should be relatively easy to customise.

Additionally, at that point of time there was no way in SwiftUI to preserve the UI state while navigating. E.g. the scrolling state was reset when you go back from the List screen to the Main screen. So the `RouterView` was just displaying all children in a `ZStack`, one on top another. Because previous children was not leaving the composition, there state of the screens was kept. Probably not a perfect way, but it worked. Maybe now there are better ways?

I also have little experience with iOS development, so I it would be really good if the community investigated different approaches and provide some feedback! The NavigationStack library looks interesting from the first sight, so worth to check if it can be integrated with Decompose.

**Comment by plusmobileapps** [comment]
Continuing conversation on the [fork](https://github.com/arkivanov/Decompose/discussions/21)
